gpytorch/lazy/triangular_lazy_tensor.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+
+from typing import Callable, Optional, Tuple, Union
+
+import torch
+from torch import Tensor
+
+from ..utils.broadcasting import _mul_broadcast_shape
+from ..utils.errors import NotPSDError
+from ..utils.memoize import cached
+from .batch_repeat_lazy_tensor import BatchRepeatLazyTensor
+from .lazy_tensor import LazyTensor
+from .non_lazy_tensor import NonLazyTensor
+
+Allsor = Union[Tensor, LazyTensor]
+
+
+class TriangularLazyTensor(LazyTensor):
+    def __init__(self, tensor: Allsor, upper: bool = False) -> None:
+        """
+        Triangular lazy tensor. Supports arbitrary batch sizes.
+
+        Args:
+            :attr:`tensor` (Tensor or LazyTensor):
+                A `b1 x ... x bk x n x n` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` triangular matrices.
+            :attr:`upper` (bool):
+                If True, the tensor is considered to be upper-triangular, otherwise lower-triangular.
+        """
+        if isinstance(tensor, TriangularLazyTensor):
+            # this is a null-op, we can just use underlying tensor directly.
+            tensor = tensor._tensor
+        elif isinstance(tensor, BatchRepeatLazyTensor):
```

Comment:
This is basically my main concern. If BatchRepeatLT.cholesky returns a TriangularLT but we didn't have this, everything would become num_repeats more expensive than necessary. We need to make sure that the new LT doesn't destroy structure when we return it from Cholesky calls. We had this bug with MulLazyTensor once and the entire package was like 10x slower

Reply:
I made sure to also add an `_expand_batch` method to TLT so that we don't unnecessarily wrap TLT in BRLTs.